File: Top 100 Codes/06-Advanced Codes related to Arrays/26-Three way partitioning of an array around a given value/Threewaypartitioningofanarrayaroundagivenvalue.py
def three_way_partition(arr, lowVal, highVal):
    start, end = 0, len(arr) - 1
    i = 0

    while i <= end:
        if arr[i] < lowVal:
            arr[i], arr[start] = arr[start], arr[i]
            start += 1
            i += 1
        elif arr[i] > highVal:
            arr[i], arr[end] = arr[end], arr[i]
            end -= 1
        else:
            i += 1

# Example usage
arr = [1, 14, 5, 20, 4, 2, 54, 20, 87, 98, 3, 1, 32]
lowVal, highVal = 14, 20
three_way_partition(arr, lowVal, highVal)
print("Partitioned array:", arr)

File: Top 100 Codes/06-Advanced Codes related to Arrays/26-Three way partitioning of an array around a given value/Threewaypartitioningofanarrayaroundagivenvalue.cpp
#include <iostream>
#include <vector>
using namespace std;

void threeWayPartition(vector<int>& arr, int lowVal, int highVal) {
    int start = 0, end = arr.size() - 1;
    int i = 0;

    while (i <= end) {
        if (arr[i] < lowVal) {
            swap(arr[i], arr[start]);
            start++;
            i++;
        }
        else if (arr[i] > highVal) {
            swap(arr[i], arr[end]);
            end--;
        }
        else {
            i++;
        }
    }
}

int main() {
    vector<int> arr = {1, 14, 5, 20, 4, 2, 54, 20, 87, 98, 3, 1, 32};
    int lowVal = 14, highVal = 20;
    threeWayPartition(arr, lowVal, highVal);

    for (int x : arr) {
        cout << x << " ";
    }
    return 0;
}
